File: sdk/java/src/main/java/dev/spark/Spark.java
package dev.spark;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import okhttp3.*;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.Map;
import java.util.concurrent.TimeUnit;

public class Spark {
    private static final String DEFAULT_BASE_URL = "https://api.spark.dev/api/v1";

    private final OkHttpClient client;
    private final Gson gson;
    private final String baseUrl;
    private String accessToken;

    public Spark() {
        this(DEFAULT_BASE_URL, null);
    }

    public Spark(String baseUrl, String accessToken) {
        this.baseUrl = baseUrl;
        this.accessToken = accessToken;
        this.gson = new Gson();
        this.client = new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .build();
    }

    public void setAccessToken(String token) {
        this.accessToken = token;
    }

    private Request.Builder auth() {
        Request.Builder builder = new Request.Builder();
        if (accessToken != null) {
            builder.header("Authorization", "Bearer " + accessToken);
        }
        builder.header("Content-Type", "application/json");
        return builder;
    }

    private <T> T request(String method, String path, Object body, Type type) throws IOException {
        String url = baseUrl + path;
        Request.Builder builder = auth().url(url);

        if (body != null) {
            String json = gson.toJson(body);
            builder.method(method, RequestBody.create(json, MediaType.get("application/json")));
        } else {
            builder.method(method, null);
        }

        try (Response response = client.newCall(builder.build()).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("HTTP " + response.code() + ": " + response.body().string());
            }
            String responseBody = response.body().string();
            return gson.fromJson(responseBody, type);
        }
    }

    public AuthResponse register(String email, String username, String password) throws IOException {
        Type type = new TypeToken<AuthResponse>() {}.getType();
        return request("POST", "/auth/register",
                Map.of("email", email, "username", username, "password", password), type);
    }

    public AuthResponse login(String email, String password) throws IOException {
        Type type = new TypeToken<AuthResponse>() {}.getType();
        AuthResponse result = request("POST", "/auth/login",
                Map.of("email", email, "password", password), type);
        if (result.accessToken != null) {
            this.accessToken = result.accessToken;
        }
        return result;
    }

    public User me() throws IOException {
        Type type = new TypeToken<User>() {}.getType();
        return request("GET", "/users/me", null, type);
    }

    public static class User {
        public String id;
        public String email;
        public String username;
        public String displayName;
        public String avatarUrl;
        public boolean isCreator;
        public boolean isVerified;
        public String createdAt;
    }

    public static class AuthResponse {
        public User user;
        public String accessToken;
        public String refreshToken;
        public int expiresIn;
    }
}

File: sdk/kotlin/src/main/kotlin/dev/spark/Spark.kt
package dev.spark

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.http.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class User(
    val id: String,
    val email: String,
    val username: String,
    @kotlinx.serialization.SerialName("display_name") val displayName: String? = null,
    @kotlinx.serialization.SerialName("avatar_url") val avatarUrl: String? = null,
    @kotlinx.serialization.SerialName("is_creator") val isCreator: Boolean = false,
    @kotlinx.serialization.SerialName("is_verified") val isVerified: Boolean = false,
    @kotlinx.serialization.SerialName("created_at") val createdAt: String,
)

@Serializable
data class AuthResponse(
    val user: User,
    @kotlinx.serialization.SerialName("access_token") val accessToken: String,
    @kotlinx.serialization.SerialName("refresh_token") val refreshToken: String,
    @kotlinx.serialization.SerialName("expires_in") val expiresIn: Int,
)

class Spark(
    private val baseUrl: String = "https://api.spark.dev/api/v1",
    accessToken: String? = null,
) {
    private var token: String? = accessToken

    private val client = HttpClient {
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true })
        }
    }

    fun setAccessToken(token: String) {
        this.token = token
    }

    private suspend fun <T> request(
        method: HttpMethod,
        path: String,
        body: Any? = null,
    ): T {
        val response = client.request("$baseUrl$path") {
            this.method = method
            token?.let { header(HttpHeaders.Authorization, "Bearer $it") }
            contentType(ContentType.Application.Json)
            if (body != null) {
                setBody(body)
            }
        }
        return response.body()
    }

    suspend fun register(email: String, username: String, password: String): AuthResponse {
        val result: AuthResponse = request(
            HttpMethod.Post,
            "/auth/register",
            mapOf("email" to email, "username" to username, "password" to password),
        )
        token = result.accessToken
        return result
    }

    suspend fun login(email: String, password: String): AuthResponse {
        val result: AuthResponse = request(
            HttpMethod.Post,
            "/auth/login",
            mapOf("email" to email, "password" to password),
        )
        token = result.accessToken
        return result
    }

    suspend fun me(): User {
        return request(HttpMethod.Get, "/users/me")
    }

    fun close() {
        client.close()
    }
}
